node/src/main/scala/org/bitcoins/node/PeerFinder.scala

```diff
@@ -0,0 +1,257 @@
+  private val _peerData: mutable.Map[Peer, PeerData] = mutable.Map.empty
+
+  private val _peersToTry: PeerStack = PeerStack()
+
+  val maxPeerSearchCount: Int = 1000
+
+  val timeoutForTrying: FiniteDuration = 12.second
+
+  private lazy val peerConnectionScheduler: Cancellable =
+    system.scheduler.scheduleWithFixedDelay(initialDelay = timeoutForTrying,
+                                            delay = timeoutForTrying) {
+      new Runnable() {
+        override def run(): Unit = {
+          logger.debug(s"Cache size: ${_peerData.size}. ${_peerData.keys}")
+          if (_peersToTry.size < 32)
+            _peersToTry.pushAll(getPeersFromDnsSeeds)
+
+          val peers = (for { _ <- 1 to 32 } yield _peersToTry.pop()).distinct
+            .filterNot(skipPeers().contains(_))
+
+          logger.debug(s"Trying next set of peers $peers")
+          peers.foreach(tryPeer)
+        }
+      }
+    }
+
+  def start: Future[Unit] = {
+    logger.debug(s"Starting PeerFinder")
+
+    (getPeersFromParam ++ getPeersFromConfig).distinct.foreach(tryPeer)
+
+    if (nodeAppConfig.enablePeerDiscovery) {
+      for {
+        (dbNonCf, dbCf) <- getPeersFromDb
+      } yield {
+        _peersToTry.pushAll(getPeersFromDnsSeeds)
+        _peersToTry.pushAll(getPeersFromResources)
+        _peersToTry.pushAll(dbNonCf)
+        _peersToTry.pushAll(dbCf, priority = 1)
+        peerConnectionScheduler //start scheduler
+        ()
+      }
+    } else {
+      logger.debug("Peer discovery disabled")
+      Future.unit
+    }
+  }
+
+  def stop: Future[Unit] = {
+    //stop scheduler
+    peerConnectionScheduler.cancel()
+    //delete try queue
+    _peersToTry.clear()
+
+    _peerData.foreach(_._2.client.close())
+
+    AsyncUtil.retryUntilSatisfied(_peerData.isEmpty,
+                                  interval = 1.seconds,
+                                  maxTries = 10)
+  }
+
+  /** creates and initialises a new test peer */
+  def tryPeer(peer: Peer): Unit = {
+    _peerData.put(peer, PeerData(peer, node))
+    _peerData(peer).peerMessageSender.connect()
+  }
+
+  def removePeer(peer: Peer): Unit = {
+    logger.debug(s"Removing peer $peer")
+    _peerData.remove(peer)
+    ()
+  }
+
+  def setServiceIdentifier(
+      peer: Peer,
+      serviceIdentifier: ServiceIdentifier): Unit = {
+    _peerData(peer).setServiceIdentifier(serviceIdentifier)
+  }
+
+  def popFromCache(peer: Peer): Option[PeerData] = {
+    if (_peerData.contains(peer))
+      _peerData.remove(peer)
+    else {
+      logger.debug(s"removeFromCache: $peer not found in peerData")
+      None
+    }
+  }
+
+  def hasPeer(peer: Peer): Boolean = {
+    _peerData.contains(peer)
+  }
+
+  def getData(peer: Peer): PeerData = {
+    assert(hasPeer(peer), "finder.getData called without any data on peer")
+    _peerData(peer)
+  }
+
+  def addToTry(peers: Vector[Peer], priority: Int = 0): Unit = {
+    _peersToTry.pushAll(peers, priority)
+  }
+}
+
+case class PeerStack() {
```

Comment:
I wanted a stack like structure that pops based on priority and then recency. PriorityQueue would have been good but needed to remove from the least priority ones too in case a higher priority values is pushed and limit is reached. So settled for a set finally.